File: Example/PINRemoteImage/ProcessingViewController.m
//
//  ProcessingViewController.m
//  PINRemoteImage
//
//  Created by Garrett Moon on 7/16/15.
//  Copyright (c) 2015 Garrett Moon. All rights reserved.
//

#import "ProcessingViewController.h"

#import <PINRemoteImage/PINImageView+PINRemoteImage.h>

@interface ProcessingViewController ()

@end

@implementation ProcessingViewController

- (void)viewDidLoad
{
    [super viewDidLoad];
    // Do any additional setup after loading the view.
}

- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];
    
    [self.imageView pin_setImageFromURL:[NSURL URLWithString:@"https://s-media-cache-ak0.pinimg.com/736x/5b/c6/c5/5bc6c5387ff6f104fd642f2b375efba3.jpg"] processorKey:@"rounded" processor:^UIImage *(PINRemoteImageManagerResult *result, NSUInteger *cost)
     {
         UIImage *image = result.image;
         CGFloat radius = 7.0f;
         CGSize targetSize = CGSizeMake(200, 300);
         CGRect imageRect = CGRectMake(0, 0, targetSize.width, targetSize.height);
         UIGraphicsBeginImageContext(imageRect.size);
         UIBezierPath *bezierPath = [UIBezierPath bezierPathWithRoundedRect:imageRect cornerRadius:radius];
         [bezierPath addClip];
         
         CGFloat widthMultiplier = targetSize.width / image.size.width;
         CGFloat heightMultiplier = targetSize.height / image.size.height;
         CGFloat sizeMultiplier = MAX(widthMultiplier, heightMultiplier);
         
         CGRect drawRect = CGRectMake(0, 0, image.size.width * sizeMultiplier, image.size.height * sizeMultiplier);
         if (CGRectGetMaxX(drawRect) > CGRectGetMaxX(imageRect)) {
             drawRect.origin.x -= (CGRectGetMaxX(drawRect) - CGRectGetMaxX(imageRect)) / 2.0;
         }
         if (CGRectGetMaxY(drawRect) > CGRectGetMaxY(imageRect)) {
             drawRect.origin.y -= (CGRectGetMaxY(drawRect) - CGRectGetMaxY(imageRect)) / 2.0;
         }
         
         [image drawInRect:drawRect];
         
         [[UIColor redColor] setStroke];
         [bezierPath setLineWidth:5.0];
         [bezierPath stroke];
        
         CGContextRef ctx = UIGraphicsGetCurrentContext();
         CGContextSetBlendMode(ctx, kCGBlendModeOverlay);
         CGContextSetAlpha(ctx, 0.5);
         
         UIImage *logo = [UIImage imageNamed:@"white-pinterest-logo"];
         CGContextScaleCTM(ctx, 1.0, -1.0);
         CGContextTranslateCTM(ctx, 0.0, -drawRect.size.height);
         CGContextDrawImage(ctx, CGRectMake(90, 10, logo.size.width, logo.size.height), [logo CGImage]);
         
         UIImage *processedImage = UIGraphicsGetImageFromCurrentImageContext();
         UIGraphicsEndImageContext();
         return processedImage;
     }];
}

/*
#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
}
*/

@end

File: Example-tvOS/PINRemoteImage.tvOSExample/ProcessingViewController.swift
//
//  ProcessingViewController.swift
//  PINRemoteImage.tvOSExample
//
//  Created by Isaac Overacker on 2/6/16.
//
//

import UIKit
import PINRemoteImage

class ProcessingViewController: UIViewController {

    @IBOutlet weak var imageView: UIImageView!

    override func viewDidLoad() {
        super.viewDidLoad()
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }

    override func viewWillAppear(animated: Bool) {
        super.viewWillAppear(animated)

        if let url = NSURL(string: "https://s-media-cache-ak0.pinimg.com/1200x/2e/0c/c5/2e0cc5d86e7b7cd42af225c29f21c37f.jpg") {
            imageView.pin_setImageFromURL(url,
                processorKey: "rounded",
                processor: { (result, cost) -> UIImage? in
                    if let image = result.image {
                        let radius = CGFloat(7.0)
                        let targetSize = CGSize(width: 200, height: 300)
                        let imageRect = CGRect(x: 0, y: 0, width: targetSize.width, height: targetSize.height)
                        UIGraphicsBeginImageContext(imageRect.size)
                        let bezierPath = UIBezierPath(roundedRect: imageRect, cornerRadius: radius)
                        bezierPath.addClip()

                        let widthMultiplier = CGFloat(targetSize.width / image.size.width)
                        let heightMultiplier = CGFloat(targetSize.height / image.size.height)
                        let sizeMultiplier = max(widthMultiplier, heightMultiplier)

                        var drawRect = CGRect(x: 0, y: 0, width: image.size.width * sizeMultiplier, height: image.size.height * sizeMultiplier)
                        if CGRectGetMaxX(drawRect) > CGRectGetMaxX(imageRect) {
                            drawRect.origin.x -= (CGRectGetMaxX(drawRect) - CGRectGetMaxX(imageRect)) / 2.0;
                        }
                        if CGRectGetMaxY(drawRect) > CGRectGetMaxY(imageRect) {
                            drawRect.origin.y -= (CGRectGetMaxY(drawRect) - CGRectGetMaxY(imageRect)) / 2.0;
                        }

                        image.drawInRect(drawRect)

                        UIColor.redColor().setStroke()
                        bezierPath.lineWidth = 5.0
                        bezierPath.stroke()

                        let ctx = UIGraphicsGetCurrentContext();
                        CGContextSetBlendMode(ctx, .Overlay);
                        CGContextSetAlpha(ctx, 0.5);

                        if let logo = UIImage(named: "white-pinterest-logo") {
                            CGContextScaleCTM(ctx, 1.0, -1.0);
                            CGContextTranslateCTM(ctx, 0.0, -drawRect.size.height);
                            CGContextDrawImage(ctx, CGRect(x: 0, y: 0, width: logo.size.width, height: logo.size.height), logo.CGImage);
                        }

                        let processedImage = UIGraphicsGetImageFromCurrentImageContext();
                        UIGraphicsEndImageContext();
                        return processedImage;
                    }

                    return nil
            })
        }
    }
}
